test_examples.sh

```diff
@@ -20,25 +34,86 @@ function cleanup {
 }
 trap cleanup EXIT
 
+export ALL_OPTS=$TEMPDIR/all_opts
+if [[ "$LONG" == "--long" ]]; then
+  p() { [ $# -eq 0 ] && echo || (shift; p "$@") | while read r ; do echo -e "$1,$r\n$r"; done }
+  echo -e $OPTS | p `cat` | sort | uniq | sed 's/,$//' | tail -n +2 > $ALL_OPTS
+else
+  echo "all" > $ALL_OPTS
+fi
+
+
 # Test file $1
 function runtest {
-  echo "running test $1"
-  yes 0 | $SOURIR "$1" --quiet --opt all > $TEMPDIR/$1.opt.out && \
-  yes 0 | $SOURIR "$1" --quiet > $TEMPDIR/$1.out && \
-  diff $TEMPDIR/$1.out $TEMPDIR/$1.opt.out > /dev/null && \
-  yes 1 | $SOURIR "$1" --quiet --opt all > $TEMPDIR/$1.opt.out && \
-  yes 1 | $SOURIR "$1" --quiet > $TEMPDIR/$1.out && \
-  diff $TEMPDIR/$1.out $TEMPDIR/$1.opt.out > /dev/null
-}
+  TEST=$1
+  INPUT=$2
+  OPT=$3
+
+  OUT="$TEMPDIR/$TEST-$INPUT"
+  BASELINE_OUT="$OUT-base.out"
+  OPT_OUT="$OUT-$opt.out"
+
+  # echo "running test $1"
+
+  if [ ! -f $BASELINE_OUT ]; then
+    yes "$INPUT" | $SOURIR "$TEST" --quiet &> "$BASELINE_OUT"
+    if [ $? -ne 0 ]; then
+      echo "Baseline run failed on $TEST with input $INP"
+      echo " ----- LOG ----------------------------------------"
+      cat $BASELINE_OUT
+      echo " --------------------------------------------------"
+      exit 255
+    fi
+  fi
+
+  yes "$INPUT" | $SOURIR "$TEST" --opt $OPT --quiet &> "$OPT_OUT"
+  if [ $? -ne 0 ]; then
+    echo "Opt run failed on $TEST with input $INP and opts $OPT"
+    echo " ----- LOG ----------------------------------------"
+    cat $OPT_OUT
+    echo " --------------------------------------------------"
+    exit 255
+  fi
+
+  diff $BASELINE_OUT $OPT_OUT > /dev/null
 
-# Iterate over examples in directory
-ALL_OK=0
-for f in *.sou; do
-  runtest $f
-  if [[ $? -ne 0 ]]; then
-    echo "Example $f failed!"
-    ALL_OK=127
+  if [ $? -ne 0 ]; then
+    echo "Output differs on $NAME with input $INP and opts $OPT"
+    echo " ----- DIFF ---------------------------------------"
+    diff $BASELINE_OUT $OPT_OUT
+    echo " --------------------------------------------------"
+    exit 255
   fi
-done
+}
+
+function status {
+  name=$1
+  done=`wc -l < $STATUS`
+  echo -ne "\e[0K\r[${done}/${NUM_TESTS}] ${name} "
+}
+
+function runtests {
+  TEST=$0
+  status "$TEST"
+  for i in `echo -e $INPUTS`
+  do
+    while read opt
+    do
+      runtest $TEST $i $opt
+    done
+  done < $ALL_OPTS
+  echo $TEST >> $STATUS
+}
+
+export -f runtests runtest status
+
+export NUM_TESTS=`find . -name '*.sou' | wc -l`
+export STATUS="$TEMPDIR/status"
+touch $STATUS
+
+find . -name '*.sou' | xargs -n 1 -P `ncores` bash -c 'runtests $@'
```

Comment:
oh, we need it. I run much more tests now. I try more inputs and in the `--long` mode I try all combinations of optimizations.

`parallel` is a dependency that you have to install (its not lsb) so it's an obstacle. Also afaik it does not exist on osx (I evaluated this choice in the past, the code is c/p from the rir testrunner).

I can make serial execution the default and parallel behind a flag. I really prefer faster tests over the output garbling. Here are the execution times on my machine:

```
# serial

[user@work sourir]$ time bash test_examples.sh --long
[45/46] ./array.sou sou 

real	0m45.547s
user	0m24.226s
sys	0m28.565s
[user@work sourir]$ time bash test_examples.sh
[45/46] ./array.sou sou 

real	0m3.541s
user	0m1.540s
sys	0m2.514s
```

```
# parallel

[user@work sourir]$ time bash test_examples.sh --long
[42/46] ./array.sou sou 

real	0m27.286s
user	0m34.210s
sys	0m39.616s
[user@work sourir]$ time bash test_examples.sh
[42/46] ./array.sou sou 

real	0m1.735s
user	0m2.010s
sys	0m3.065s
```